modules/product/src/main/java/com/opengamma/strata/product/cms/CmsLeg.java

```diff
@@ -0,0 +1,1058 @@
+
+import org.joda.beans.Bean;
+import org.joda.beans.BeanDefinition;
+import org.joda.beans.ImmutableBean;
+import org.joda.beans.ImmutableConstructor;
+import org.joda.beans.ImmutablePreBuild;
+import org.joda.beans.JodaBeanUtils;
+import org.joda.beans.MetaProperty;
+import org.joda.beans.Property;
+import org.joda.beans.PropertyDefinition;
+import org.joda.beans.impl.direct.DirectFieldsBeanBuilder;
+import org.joda.beans.impl.direct.DirectMetaBean;
+import org.joda.beans.impl.direct.DirectMetaProperty;
+import org.joda.beans.impl.direct.DirectMetaPropertyMap;
+
+import com.google.common.collect.ImmutableList;
+import com.opengamma.strata.basics.PayReceive;
+import com.opengamma.strata.basics.currency.Currency;
+import com.opengamma.strata.basics.date.DayCount;
+import com.opengamma.strata.basics.date.DaysAdjustment;
+import com.opengamma.strata.basics.schedule.PeriodicSchedule;
+import com.opengamma.strata.basics.schedule.Schedule;
+import com.opengamma.strata.basics.schedule.SchedulePeriod;
+import com.opengamma.strata.basics.value.ValueSchedule;
+import com.opengamma.strata.collect.ArgChecker;
+import com.opengamma.strata.product.swap.SwapIndex;
+import com.opengamma.strata.product.swap.type.IborRateSwapLegConvention;
+
+/**
+ * A CMS leg of a constant maturity swap (CMS) product. 
+ * <p>
+ * The class defines a single CMS leg for CMS or CMS cap/floor. 
+ * The CMS leg of the swap periodically pays coupons based on swap rate, the observed value of {@linkplain SwapIndex swap index}, 
+ * whereas that of CMS cap/floor is a set of call/put options on successive swap rates, i.e., CMS caplets/floorlets.
+ * <p>
+ * The periodic payments generated by {@link #expand()} are CMS coupons, CMS caplets or CMS floorlets following the 
+ * rule below.
+ * <p>
+ * {@code capSchedule} is used to represent strike values of individual caplets, whereas {@code floorSchedule} is used 
+ * to represent strike values of individual floorlets. 
+ * Thus at least one of {@code capSchedule} and {@code floorSchedule} must be empty. 
+ * If both the fields are null, the periodic payments in this leg are CMS coupons. 
+ */
+@BeanDefinition
+public final class CmsLeg
+    implements ImmutableBean, Serializable {
+
+  /**
+   * Whether the leg is pay or receive.
+   * <p>
+   * A value of 'Pay' implies that the resulting amount is paid to the counterparty.
+   * A value of 'Receive' implies that the resulting amount is received from the counterparty.
+   * Note that negative swap rates can result in a payment in the opposite direction to that implied by this indicator.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final PayReceive payReceive;
+  /**
+   * The periodic payment schedule.
+   * <p>
+   * This is used to define the periodic payment periods.
+   * These are used directly or indirectly to determine other dates in the leg.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final PeriodicSchedule periodicSchedule;
```

Comment:
Are stubs valid? If not, then should validate that. If they are, then should document that.